### PR TITLE
BUG: pandas Timestamp tz_localize and tz_convert do not preserve `freq` attribute

### LINKED ISSUE #25241: BUG: pandas Timestamp tz_localize and tz_convert do not preserve `freq` attribute
The `pandas.Timestamp.tz_localize()` and `pandas.Timestamp.tz_convert()` methods do not preserve the `freq` attribute.  For background, the `freq` attribute is mostly a "helpful label", and it seems semantically ill-defined to me; cf. issue #15146, which proposes removing the attribute.

#### Code Sample

```
>>> import pandas
>>> import pytz

>>> t1 = pandas.Timestamp('2019-01-01 10:00', freq='H')
>>> print(t1.tz_localize(pytz.utc).freq)  
None

>>> t2 = pandas.Timestamp('2019-01-02 12:00', tz=pytz.utc, freq='T')
>>> print(t2.tz_convert(pytz.utc).freq)
None
```

#### Problem description
I don't think that `pandas.Timestamp` should actually have a `freq` attribute, but so long as it does, it should be propagated to the new timestamp value returned by `tz_localize()` or `tz_convert`.  In the `tz_convert` case, the method should produce an "equivalent" timestamp with a different tz.

#### Expected Output

`t1.tz_localize(pytz.utc).freq` should equal `t1.freq`, and `t2.tz_convert(pytz.utc).freq` should equal `t2.freq`.

#### Output of ``pd.show_versions()``

<details>

No module named 'dask'

INSTALLED VERSIONS
------------------
commit: 04df22fbeec782f474bcff20480eded04e0d1d82
python: 3.7.2.final.0
python-bits: 64
OS: Darwin
OS-release: 17.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.25.0.dev0+89.g04df22fbe
pytest: 4.2.0
pip: 19.0.1
setuptools: 40.7.3
Cython: 0.29.4
numpy: 1.15.4
scipy: 1.2.0
pyarrow: 0.11.1
xarray: 0.11.0
IPython: 7.2.0
sphinx: 1.8.4
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.9
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.14
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml.etree: 4.3.0
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: 1.2.17
pymysql: 0.9.3
psycopg2: None
jinja2: 2.10
s3fs: 0.2.0
fastparquet: 0.2.1
pandas_gbq: None
pandas_datareader: None
gcsfs: 0.1.2

</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [ ] closes #25241
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/tslibs/test_conversion.py [hunk 1]
```diff
@@ -66,3 +66,12 @@ def test_length_zero_copy(dtype, copy):
...
     assert result.base is (None if copy else arr)
+
+
+def test_tz_convert_freq():
+    import pandas as pd
+    import pytz
+    t1 = pd.Timestamp('2019-01-01 10:00', freq='H')
+    assert t1.tz_localize(pytz.utc).freq == t1.freq
+    t2 = pd.Timestamp('2019-01-02 12:00', tz=pytz.utc, freq='T')
+    assert t2.tz_convert(pytz.utc).freq == t2.freq
```
Comment (nmusolino) [inline pandas/tests/tslibs/test_conversion.py]: These four lines look nice.

I think these tests would fit better in `pandas/tests/scalar/timestamp/test_timestamp.py`, which already contains tests of `Timestamp.tz_localize` and `Timestamp.tz_convert`.

**Review by nmusolino — CHANGES_REQUESTED** [review]

File: pandas/tests/tslibs/test_conversion.py [hunk 2]
```diff
@@ -66,3 +66,12 @@ def test_length_zero_copy(dtype, copy):
     arr = np.array([], dtype=dtype)
     result = conversion.ensure_datetime64ns(arr, copy=copy)
     assert result.base is (None if copy else arr)
+
+
+def test_tz_convert_freq():
```
Comment (mroeschke) [inline pandas/tests/tslibs/test_conversion.py]: Can you use the `tz_naive_fixture`. Ensures we also test `tz_localize(None)` as well.

File: pandas/tests/tslibs/test_conversion.py [hunk 3]
```diff
@@ -66,3 +66,12 @@ def test_length_zero_copy(dtype, copy):
...
+    import pandas as pd
+    import pytz
+    t1 = pd.Timestamp('2019-01-01 10:00', freq='H')
+    assert t1.tz_localize(pytz.utc).freq == t1.freq
```
Comment (mroeschke) [inline pandas/tests/tslibs/test_conversion.py]: Just assert the `Timestamp`s are equal. IIRC the freq will be compared as well.

File: pandas/tests/tslibs/test_conversion.py [hunk 4]
```diff
@@ -66,3 +66,12 @@ def test_length_zero_copy(dtype, copy):
     arr = np.array([], dtype=dtype)
     result = conversion.ensure_datetime64ns(arr, copy=copy)
     assert result.base is (None if copy else arr)
+
+
+def test_tz_convert_freq():
+    import pandas as pd
+    import pytz
```
Comment (mroeschke) [inline pandas/tests/tslibs/test_conversion.py]: Don't need to import this. Just the string `'UTC'` will automatically use `pytz.UTC`

**Review by mroeschke — COMMENTED** [review]
Needs a whatsnew for 0.25.0

File: pandas/tests/tslibs/test_conversion.py (hunk 3 above)
Comment (charlesdong1991) [reply pandas/tests/tslibs/test_conversion.py]: but i assume `t1.tz_localize(pytz.utc)` is `tz-aware` while `t1` is `tz-naive` ? @mroeschke

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 5]
```diff
@@ -780,6 +780,13 @@ def test_hash_equivalent(self):
         stamp = Timestamp(datetime(2011, 1, 1))
         assert d[stamp] == 5
 
+    def test_tz_convert_freq(self, tz_naive_fixture):
+        # GH25241
+        t1 = Timestamp('2019-01-01 10:00', freq='H')
+        assert t1.tz_localize(tz=tz_naive_fixture).freq == t1.freq
+        t2 = Timestamp('2019-01-02 12:00', tz='UTC', freq='T')
+        assert t2.tz_convert(tz='UTC') == t2
```
Comment (nmusolino) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: I think you need to assert that the `freq` attributes are equal, as opposed to comparing the timestamps.

Timestamp `freq` values do not influence [equality comparisons](https://github.com/pandas-dev/pandas/blob/2448e5229683acbe7de57f2d53065247aa085b1f/pandas/_libs/tslibs/timestamps.pyx#L216) or hash values.  Here's a quick example that shows this:

```
>>> t1 = pandas.Timestamp('2018-01-01 10:00', freq='H') 
>>> t2 = pandas.Timestamp('2018-01-01 10:00', freq='T') 
>>> t1 == t2
True
>>> t1.freq == t2.freq
False
```

**Review by nmusolino — CHANGES_REQUESTED** [review]

File: pandas/tests/scalar/timestamp/test_timestamp.py (hunk 5 above)
Comment (charlesdong1991) [reply pandas/tests/scalar/timestamp/test_timestamp.py]: ahh, nice catch!! 
```python
t1 = pd.Timestamp('2018-01-01 10:00', freq='H')
t2 = pd.Timestamp('2018-01-01 10:00')])
```
and `t1 != t2`! thanks! @nmusolino

File: pandas/tests/tslibs/test_conversion.py (hunk 3 above)
Comment (mroeschke) [reply pandas/tests/tslibs/test_conversion.py]: I guess as mentioned below. I was incorrect in assuming the freq would be compared as well.

File: doc/source/whatsnew/v0.25.0.rst [hunk 6]
```diff
@@ -93,7 +93,7 @@ Timezones
 ^^^^^^^^^
 
 - Bug in :func:`to_datetime` with ``utc=True`` and datetime strings that would apply previously parsed UTC offsets to subsequent arguments (:issue:`24992`)
--
+- Bug in :func:`tz_localize` and :func:`tz_convert` does not propagate freq if ``Timestamp`` is initiated with freq (:issue:`25241`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v0.25.0.rst]: Could you specify `Timestamp.tz_localize` and `Timestamp.tz_convert`? Then you can remove the additional ``Timestamp`` towards the end.

Double back ticks around ``freq`` as well.

**Review by mroeschke — CHANGES_REQUESTED** [review]
Looks pretty good. One small whatsnew comment.

Could you also test that this doesn't happen with `DatetimeIndex.tz_localize` and `DatetimeIndex.tz_convert`?

File: doc/source/whatsnew/v0.25.0.rst (hunk 6 above)
Comment (charlesdong1991) [reply doc/source/whatsnew/v0.25.0.rst]: done, thanks for your review! @mroeschke

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 7]
```diff
@@ -780,6 +780,19 @@ def test_hash_equivalent(self):
...
 
+    def test_tz_conversion_freq(self, tz_naive_fixture):
+        # GH25241
+        t1 = Timestamp('2019-01-01 10:00', freq='H')
+        assert t1.tz_localize(tz=tz_naive_fixture).freq == t1.freq
+        t2 = Timestamp('2019-01-02 12:00', tz='UTC', freq='T')
+        assert t2.tz_convert(tz='UTC').freq == t2.freq
+
+        # freq propagation error doesn't happen with DatetimeIndex
+        t3 = DatetimeIndex(['2019-01-01 10:00'], freq='H')
```
Comment (mroeschke) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: These are better suited in `pandas/tests/indexes/datetimes/test_timezones.py`. After moved, LGTM

Comment (charlesdong1991) [reply pandas/tests/scalar/timestamp/test_timestamp.py]: moved! thanks! @mroeschke

**Review by mroeschke — APPROVED** [review]

**Review by nmusolino — APPROVED** [review]
Looks good to me, too.

**Comment by jreback** [comment]
thanks @charlesdong1991 

I agree we should simply deprecate freq on a Timestamp though

**Pull request merged**